pkg/deploy/lattice/listener_manager.go

```diff
@@ -45,45 +47,56 @@ func (d *defaultListenerManager) Upsert(
 	ctx context.Context,
 	modelListener *model.Listener,
 	modelSvc *model.Service,
+	defaultAction *vpclattice.RuleAction,
 ) (model.ListenerStatus, error) {
 	if modelSvc.Status == nil || modelSvc.Status.Id == "" {
 		return model.ListenerStatus{}, errors.New("model service is missing id")
 	}
 
 	d.log.Infof("Upsert listener %s-%s", modelListener.Spec.K8SRouteName, modelListener.Spec.K8SRouteNamespace)
-
-	latticeListener, err := d.findListenerByPort(ctx, modelSvc.Status.Id, modelListener.Spec.Port)
+	latticeSvcId := modelSvc.Status.Id
+	latticeListenerSummary, err := d.findListenerByPort(ctx, latticeSvcId, modelListener.Spec.Port)
 	if err != nil {
 		return model.ListenerStatus{}, err
 	}
-	if latticeListener != nil {
-		// we do not support listener updates as the only mutable property
-		// is the default action, which we set to 404 as required by the gw spec
-		// so here we just return the existing one
-		d.log.Debugf("Found existing listener %s, nothing to update", aws.StringValue(latticeListener.Id))
-		return model.ListenerStatus{
-			Name:        aws.StringValue(latticeListener.Name),
-			ListenerArn: aws.StringValue(latticeListener.Arn),
-			Id:          aws.StringValue(latticeListener.Id),
-			ServiceId:   modelSvc.Status.Id,
-		}, nil
+	if latticeListenerSummary == nil {
+		// listener not found, create new one
+		return d.create(ctx, latticeSvcId, modelListener, defaultAction)
 	}
 
-	// no listener currently exists, create
-	defaultStatus := aws.Int64(404)
-	defaultResp := vpclattice.FixedResponseAction{
-		StatusCode: defaultStatus,
+	existingListenerStatus := model.ListenerStatus{
+		Name:        aws.StringValue(latticeListenerSummary.Name),
+		ListenerArn: aws.StringValue(latticeListenerSummary.Arn),
+		Id:          aws.StringValue(latticeListenerSummary.Id),
+		ServiceId:   latticeSvcId,
+	}
+	if modelListener.Spec.Protocol != vpclattice.ListenerProtocolTlsPassthrough {
```

Comment:
With the refactoring proposed, you should be able to simply compare the default action on the model vs actual and we can drop any special handling for different protocol types here.

Reply:
My intention here is to save one api call for `vpclattice.GetListener()` for non-TlsPassthrough listener, not necessary?

Reply:
OK. I think this is fine as-is.